src/sim/detail/InitializeTracks.cu

```diff
@@ -0,0 +1,339 @@
+ * that survived cutoffs for each interaction.
+ */
+void locate_alive(StatePointers            states,
+                  ParamPointers            params,
+                  TrackInitializerPointers inits)
+{
+    KernelParamCalculator calc_launch_params;
+    auto                  lparams = calc_launch_params(states.size());
+    locate_alive_kernel<<<lparams.grid_size, lparams.block_size>>>(
+        states, params, inits);
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create track initializers from primary particles.
+ */
+void process_primaries(span<const Primary>      primaries,
+                       TrackInitializerPointers inits)
+{
+    KernelParamCalculator calc_launch_params;
+    auto                  lparams = calc_launch_params(primaries.size());
+    process_primaries_kernel<<<lparams.grid_size, lparams.block_size>>>(
+        primaries, inits);
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create track initializers from secondary particles.
+ */
+void process_secondaries(StatePointers            states,
+                         ParamPointers            params,
+                         TrackInitializerPointers inits)
+{
+    KernelParamCalculator calc_launch_params;
+    auto                  lparams = calc_launch_params(states.size());
+    process_secondaries_kernel<<<lparams.grid_size, lparams.block_size>>>(
+        states, params, inits);
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Remove all elements in the vacancy vector that were flagged as active
+ * tracks.
+ */
+size_type remove_if_alive(span<size_type> vacancies)
+{
+    thrust::device_ptr<size_type> end = thrust::remove_if(
+        thrust::device_pointer_cast(vacancies.data()),
+        thrust::device_pointer_cast(vacancies.data() + vacancies.size()),
+        IsEqual{flag_alive()});
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+
+    // New size of the vacancy vector
+    size_type result = thrust::raw_pointer_cast(end) - vacancies.data();
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sum the total number of surviving secondaries.
+ */
+size_type reduce_counts(span<size_type> counts)
+{
+    size_type result = thrust::reduce(
+        thrust::device_pointer_cast(counts.data()),
+        thrust::device_pointer_cast(counts.data()) + counts.size(),
+        size_type(0),
+        thrust::plus<size_type>());
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the exclusive prefix sum of the number of surviving secondaries
+ * from each interaction.
```

Comment:
Can you add a line explaining "exclusive prefix sum"? I think it's doing the same as `numpy.cumsum(a, out=a)`?

Reply:
Yes, I'll add a better explanation of what this is doing. This would be equivalent to `np.cumsum(np.concatenate(([0], a[:-1])), out=a)`, so if the input were `a = [2, 4, 1, 0, 3]`, the result would be `a = [0, 2, 6, 7, 7]`. I.e., when calculating the output at index `i`, the `i`'th element isn't included in the sum.